src/opm/parser/eclipse/EclipseState/Schedule/Well/WellTestState.cpp

```diff
@@ -19,72 +19,95 @@
 #include <stdexcept>
 #include <algorithm>
 
+#include <cassert>
 #include <opm/parser/eclipse/EclipseState/Schedule/Well/WellTestConfig.hpp>
 #include <opm/parser/eclipse/EclipseState/Schedule/Well/WellTestState.hpp>
+#include <opm/common/OpmLog/OpmLog.hpp>
 
 namespace Opm {
 
-    void WellTestState::addClosedWell(const std::string& well_name, WellTestConfig::Reason reason, double sim_time) {
-        if (this->hasWell(well_name, reason))
-            return;
+    void WellTestState::closeWell(const std::string& well_name, WellTestConfig::Reason reason, double sim_time) {
+
+        WTestWell* well_ptr = getWell(well_name, reason);
 
-        this->wells.push_back({well_name, reason, sim_time, 0});
+        if (well_ptr) {
+            // the well exists already, we just update it with action of closing
+            well_ptr->closed = true;
+            well_ptr->last_test = sim_time;
+        } else
+            this->wells.push_back({well_name, reason, true, sim_time, 0});
     }
 
 
-    void WellTestState::openWell(const std::string& well_name) {
-        wells.erase(std::remove_if(wells.begin(),
-                                   wells.end(),
-                                   [&well_name](const ClosedWell& well) { return (well.name == well_name); }),
-                    wells.end());
+    void WellTestState::openWell(const std::string& well_name, WellTestConfig::Reason reason) {
+
+        WTestWell* well_ptr = getWell(well_name, reason);
+
+        if (well_ptr)
+            well_ptr->closed = false;
+        else
+            throw std::runtime_error("No well named " + well_name + " with close reason "
+                                    + WellTestConfig::reasonToString(reason)
+                                    + " found in WellTestState.");
     }
 
 
     void WellTestState::dropWell(const std::string& well_name, WellTestConfig::Reason reason) {
         wells.erase(std::remove_if(wells.begin(),
                                    wells.end(),
-                                   [&well_name, reason](const ClosedWell& well) { return (well.name == well_name && well.reason == reason); }),
+                                   [&well_name, reason](const WTestWell& well) { return (well.name == well_name && well.reason == reason); }),
```

Comment:
I think the `dropWell()` method can be removed now?

Reply:
Sure.